File: code/3.longest-substring-without-repeating-characters.ts
function lengthOfLongestSubstring(s: string): number {
    // index of last occurrence of a character
    let lastOccurrence = new Map<string, number>();
    let longest = 0;
    let start = 0;

    for (let end = 0; end < s.length; end++) {
        if (lastOccurrence.has(s[end])) {
            const currLastOccurrence = lastOccurrence.get(s[end]);
            if (currLastOccurrence + 1 > start) start = currLastOccurrence + 1;
        }

        lastOccurrence.set(s[end], end);

        const currSubstringLength = end + 1 - start;
        if (currSubstringLength > longest) longest = currSubstringLength;
    }

    return longest;
}

File: code/3.longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
	// last[b] = last (1-based) index byte b was seen at in s
	// alternatively: the index after the last occurence of b in s
	var last [256]int
	longest, start := 0, 0

	for end := range s {
		b := s[end]

		if last[b] > start {
			start = last[b]
		}

		last[b] = end + 1

		if length := end + 1 - start; length > longest {
			longest = length
		}
	}

	return longest
}
